src/hotspot/share/opto/loopUnswitch.cpp

```diff
@@ -222,69 +227,172 @@ bool PhaseIdealLoop::has_control_dependencies_from_predicates(LoopNode* head) co
+// becomes the false-path-loop while original loop becomes the true-path-loop.
+class OriginalLoop : public StackObj {
+  LoopNode* const _strip_mined_loop_head;
+  IdealLoopTree* const _loop;
+  Node_List* const _old_new;
+  PhaseIdealLoop* const _phase;
+
+ public:
+  OriginalLoop(IdealLoopTree* loop, Node_List* old_new)
+      : _strip_mined_loop_head(loop->_head->as_Loop()->skip_strip_mined()),
+        _loop(loop),
+        _old_new(old_new),
+        _phase(loop->_phase) {}
+  NONCOPYABLE(OriginalLoop);
+
+ private:
+  void fix_loop_entries(IfProjNode* true_path_loop_entry, IfProjNode* false_path_loop_entry) {
+    _phase->replace_loop_entry(_strip_mined_loop_head, true_path_loop_entry);
+    LoopNode* false_path_loop_strip_mined_head = old_to_new(_strip_mined_loop_head)->as_Loop();
+    _phase->replace_loop_entry(false_path_loop_strip_mined_head, false_path_loop_entry);
+  }
+
+  Node* old_to_new(const Node* old) const {
+    return _old_new->at(old->_idx);
+  }
+
 #ifdef ASSERT
-void PhaseIdealLoop::verify_fast_loop(LoopNode* head, const ProjNode* proj_true) const {
-  assert(proj_true->is_IfTrue(), "must be true projection");
-  Node* entry = head->skip_strip_mined()->in(LoopNode::EntryControl);
-  Predicates predicates(entry);
-  if (!predicates.has_any()) {
-    // No Parse Predicate.
-    Node* uniqc = proj_true->unique_ctrl_out();
-    assert((uniqc == head && !head->is_strip_mined()) || (uniqc == head->in(LoopNode::EntryControl)
-                                                          && head->is_strip_mined()), "must hold by construction if no predicates");
-  } else {
-    // There is at least one Parse Predicate. When skipping all predicates/predicate blocks, we should end up
-    // at 'proj_true'.
-    assert(proj_true == predicates.entry(), "must hold by construction if at least one Parse Predicate");
+  void verify_unswitched_loop_versions(LoopNode* true_path_loop_head,
+                                       const UnswitchedLoopSelector& unswitched_loop_selector) const {
+    verify_unswitched_loop_version(true_path_loop_head, unswitched_loop_selector.true_path_loop_proj());
+    verify_unswitched_loop_version(old_to_new(true_path_loop_head)->as_Loop(),
+                                   unswitched_loop_selector.false_path_loop_proj());
+  }
+
+  static void verify_unswitched_loop_version(LoopNode* loop_head, IfProjNode* loop_selector_if_proj) {
+    Node* entry = loop_head->skip_strip_mined()->in(LoopNode::EntryControl);
+    const Predicates predicates(entry);
+    // When skipping all predicates, we should end up at 'loop_selector_if_proj'.
+    assert(loop_selector_if_proj == predicates.entry(), "should end up at loop selector If");
   }
-}
 #endif // ASSERT
 
+ public:
+  // Unswitch on the invariant loop selector by creating two unswitched loop versions.
+  void unswitch(const UnswitchedLoopSelector& unswitched_loop_selector) {
```

Comment:
Just a suggestion, I am also ok with what it is now:
I wonder if the name `unswitch` is right here. Because at this point you only clone the loops, you don't yet remove the if. Maybe this should for now just be called "duplicate", and `create_unswitched_loop_versions` then really does the unswitching, by both duplicating and removing the ifs?